xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.h

```diff
@@ -38,12 +50,74 @@ class CDVDSubtitlesLibass
 
   bool DecodeHeader(char* data, int size);
   bool DecodeDemuxPkt(const char* data, int size, double start, double duration);
+
+  /*!
+  * \brief Create a new ASS track with a default ASS style (used on ASS events)
+  * \param useDebugStyle Use style text/color/position for debug info purpose
+  * \return True if success, false if error
+  */
+  bool CreateTrack(bool useDebugStyle = false);
```

Comment:
this seems hackish. The libasshandler (this class) shouldn't need to know what a debug style is. You want to abstract the style into some kodi class or struct and feed it to the libasshandler instead. The handler should then be able to do the necessary mapping or conversion to something libass understands

Reply:
this is used only from DebugRenderer only, we don't need to change libass style properties never,
i don't think it's worth spending a lot of time redefining more than 15 libass properties just to create a debug style used only by developers

Reply:
It's pretty much irrelevant since the style should be computed by the `OverlayRenderer` (or `DebugRenderer`) and passed to the libass handler instead. So you'll need to abstract the style anyway, either for the regular subs or for the debug overlay (which will be always a fixed style).

Reply:
I would rather create a new different method

Reply:
this will make the code huge and heavy to understand...

Reply:
i need to create a new style struct to pass all variables

Reply:
theorising about doing this thing, i can not move it in the `OverlayRenderer` because i need create the ASS style before add the ASS events, then when the parser is just "Open"

Reply:
Nobody said this would be easy or quick, it won't. A lot of effort has been placed into refactoring Kodi and making sure new additions do not deteriorate the architecture.

This class is an handler for libass, means it calls libass specific methods. It shouldn't be aware of styles, settings (except libass specific), resolutions, state, etc. I think creating a style struct or class is the best way to go, it allows to abstract all the available Kodi sub style options and to easily map them to the ass_style equivalent. It will also be easier to extend in the future with new style options.

Reply:
Note the style depends on the user configured settings, the user wants to change a setting and the style should follow, even during the playback. It can't be static and configured once. Libass handles styles just like events as far as I remember. You need to allocate a style and you get an id returned, you then set the required ASS_style members. You are then free to change those members during runtime if needed.
Afaics the only way to do this is to pass the style before calling into RenderImage. The component controlling all this workflow is the OverlayRenderer

Reply:
This means generating styles for each frame, with spending resources for no reason
or you have a system to intercept changes in Kodi settings (also in a generic way) and update styles only at that particular time

Reply:
The concept of settings change callbacks already exist, so thats not a concern just fyi

https://github.com/xbmc/xbmc/blob/master/xbmc/settings/lib/ISettingCallback.h

Reply:
I will take a look thanks

Reply:
Yeap, no need to feed a new style on each frame, Kodi is able to handle settings callback. Out of curiosity, when working on this I was precisely on this point, trying to group all subtitle settings into a single place just like other components (e.g. games - https://github.com/xbmc/xbmc/blob/a80d1293b0579bbed7dd7d20304ba34a20519f78/xbmc/settings/GameSettings.h) do.

Having something similar would allow any component to register as an observer (https://github.com/xbmc/xbmc/blob/a80d1293b0579bbed7dd7d20304ba34a20519f78/xbmc/utils/Observer.cpp#L32) and be notified of changes (https://github.com/xbmc/xbmc/blob/a80d1293b0579bbed7dd7d20304ba34a20519f78/xbmc/games/GameSettings.cpp#L98). The problem is that subtitle settings are pretty much everywhere and you can easily go into the rabbit hole. Either way, it's the right way to do it even though it needs a far bit of work

Reply:
done all